File: solution/python3/49.group-anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#

# @lc code=start
from collections import Counter
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = {}
        for s in strs:
            if not len(s) in d:
                d[len(s)] = {}

            if s[0] in d[len(s)]:
                d[len(s)][s[0]].append(s)
            else:
                d[len(s)][s[0]] = [s]

        ans = []
        for s in strs:
            words = d[len(s)]
            group = []
            for c in s:
                tmp = words[c][:]
                for word in tmp:
                    if Counter(s) == Counter(word):
                        words[c].remove(s)
                        group.append(s)
            ans.append(group)
        return ans
# @lc code=end

File: solution/java8/49.group-anagrams.java
/*
 * @lc app=leetcode id=49 lang=java
 *
 * [49] Group Anagrams
 */

// @lc code=start

import java.util.*;
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        int N = strs.length;
        List<Map<String, Integer>> counters = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            Map<String, Integer> counter = new HashMap<>();
            for (String s: strs[i].split("")) {
                if (counter.containsKey(s)) {
                    int tmp = counter.get(s);
                    counter.put(s, tmp + 1);
                } else {
                    counter.put(s, 1);
                }
            }
            counters.add(counter);
        }

        List<List<String>> ans = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            if (strs[i] == null) {
                continue;
            }
            List<String> words = new ArrayList<>();
            words.add(strs[i]);
            for (int j = i+1; j < N; j++) {
                if (counters.get(i).equals(counters.get(j))) {
                    words.add(strs[j]);
                    strs[j] = null;
                }
            }
            ans.add(words);
        }

        return ans;
    }
}
// @lc code=end

